Add new metric APIs

noahfalk opened the pull request:
We added new metric APIs in .NET 6 and we need to tell customers about them.
These pages should introduce the new APIs, describe how they differ from existing APIs, and provide guidance how to instrument and collect.

noahfalk committed:
[WIP] Add new metric APIs

We added new metric APIs in .NET 6 and we need to tell customers about them.
These initial pages disambiguate the API from pre-existing APIs and help
users pick.

I plan to add more content that lets devs understand and use
the new APIs in a new page that is parallel to the EventCounters page, but
I haven't gotten to that part yet.

noahfalk commented:
cc @dotnet/dotnet-diag

noahfalk commented:
cc @tarekgh

tarekgh commented:
@noahfalk would make sense to make this as draft PR till it is ready for review?

noahfalk commented:
> @noahfalk would make sense to make this as draft PR till it is ready for review?

Yeah, but I think I missed the button when I filed it and I didn't see a way to convert it to draft afterwards. Is there a way?

hoyosjs commented:
As far as I know, no. This is a one way street.

hoyosjs commented:
0.0 magic. @tarekgh Where is the option to? It used to not be possible

tarekgh commented:
@noahfalk I converted it to a draft PR. I pasted the image for the UI place to convert the PR to draft.

![image](https://user-images.githubusercontent.com/10833894/126845749-5ca16d35-6ffd-4e10-aad6-c910e31037fe.png)

hoyosjs commented on docs/core/diagnostics/metrics.md:
```diff
@@ -0,0 +1,56 @@
+1. Some .NET libraries are already instrumented with metrics by the library's authors. Check the library's
+documentation to learn what may be available.
+  - [Built-in Metrics](https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/performance-counters) in .NET Framework 4.8 and lower
+  - [Built-in Metrics](.\available-counters.md) in .NET Core 3.1, .NET 5 and .NET 6
```
```suggestion
  - [Built-in Metrics](available-counters.md) in .NET Core 3.1, .NET 5 and .NET 6
```

noahfalk commented:
Thanks @tarekgh !

gfoidl commented on docs/core/diagnostics/metrics.md:
```diff
@@ -0,0 +1,56 @@
+that want to create custom tooling or adapters to other systems.
+
+- [EventCounters](./event-counters.md) - This was .NET's first set of cross platform metrics APIs. They are available by
+targetting .NET Core 3.1 and a small subset is available on .NET Framework 4.7.1 and above. These
```
```suggestion
targetting .NET Core 3.1 or later and a small subset is available on .NET Framework 4.7.1 and above. These
```

noahfalk committed:
Finish 1st draft of metrics docs

noahfalk commented:
@tarekgh @reyang @tarekgh @cijothomas @shirhatti @hoyosjs @gfoidl  - just pinging to make sure you see the changes. I'm expecting small updates on the PR to be much more frequent now. Thanks all!

I'm still doing some revision/editing, but I think the overall structure is complete and content is filled in. If you have any suggestions big or small, now is probably a good time to take a look. I see you guys made a few comments on my early draft that I have not yet reviewed. Promise I'm not ignoring it, just haven't gotten there quite yet : )

My plan was to merge the PR no earlier than when the official RTM builds of DiagnosticSource are available because the text doesn't refer to the .NET portions as being preview builds.

@hoyosjs - I am planning to go through all that great feedback you wrote during the exploration a few months ago and cross-check if the docs adequately offer guidance for the places you found confusing. Probably there are still some gaps that I'll need to work on addressing.

cijothomas commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,452 @@
+
+# Creating Metrics
+
+**This article applies to: ✔️** .NET Core 3.1 and later versions **✔️** .NET Framework 4.5 and later versions
```
```suggestion
**This article applies to: ✔️** .NET Core 3.1 and later versions **✔️** .NET Framework 4.6.1 and later versions
```

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+
+# Collect metrics
+
+**This article applies to: ✔️** .NET Core 3.1 and later versions **✔️** .NET Framework 4.5 and later versions
```
>NET Framework 4.5  [](http://example.com/codeflow?start=73&length=18)

4.6.1

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+**Prerequisites**: [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet) or a later version
+
+Before metrics can be collected, we need to produce some measurements. For simplicity we will create a small app that has
+some trivial metric instrumentation. The .NET runtime also has [a variety of metrics of built-in](available-counters.md). See 
```
>available-counters.md [](http://example.com/codeflow?start=98&length=21)

does this file already exist?

noahfalk replied:
yeah it does

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+instrumentation is working correctly. It works with both <xref:System.Diagnostics.Metrics?displayProperty=nameWithType> based
+APIs as well as [EventCounters](event-counters.md).
+
+If the [dotnet-counters](./dotnet-counters.md) tool is not already installed use the SDK to install it:
```
>./ [](http://example.com/codeflow?start=25&length=2)

nit: why using `./` here? you didn't use it with the links.

noahfalk replied:
I haven't been consistent. My understanding is that both forms work and I think without the './' is more common elsewhere so I am converging on that.

reyang replied:
In OSS projects I've seen more `./` usage (e.g. https://github.com/open-telemetry/opentelemetry-specification/blob/main/README.md).

I guess this was coming from the Linux shell habit - `./a.out foo bar` would run but `a.out foo bar` would not.

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+If the [dotnet-counters](./dotnet-counters.md) tool is not already installed use the SDK to install it:
+
+```dotnetcli
+> dotnet tool update -g dotnet-counters
````
>dotnet tool update -g dotnet-counters [](http://example.com/codeflow?start=2&length=37)

just checking, if anyone using .NET Core 3.1 and run this command, still can get the latest version of the tool. is that right?

noahfalk replied:
Yep, I did all the testing with only the 3.1 SDK + runtime installed to confirm no later dependencies crept in:

```
> dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.414
 Commit:    88a3b0c5ca

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.414\

Host (useful for support):
  Version: 3.1.20
  Commit:  2833dab8c3

.NET Core SDKs installed:
  3.1.414 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

> dotnet tool uninstall dotnet-counters -g
Tool 'dotnet-counters' (version '5.0.251802') was successfully uninstalled.

> dotnet tool update -g dotnet-counters
You can invoke the tool using the following command: dotnet-counters
Tool 'dotnet-counters' (version '5.0.251802') was successfully installed.
```

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+```
+
+Find the ID for the process name that matches the sample app and have dotnet-counters monitor all metrics from the
+"HatCo.HatStore" Meter.
````
>"HatCo.HatStore" [](http://example.com/codeflow?start=0&length=16)

should you mentioned the meter name is case sensitive? or dotnet-counter tool handle that?

noahfalk replied:
It is a case-sensitive comparison. I added a note in the text to call this out.

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+1. Add a reference to the OpenTelemetry Prometheus exporter to the sample application:
+
+```dotnetcli
+dotnet add package OpenTelemetry.Exporter.Prometheus --version 1.2.0-beta1
````
>--version 1.2.0-beta1 [](http://example.com/codeflow?start=53&length=21)

I am not sure if it is worth it to mention can use `--prerelease` instead if using .NET 5.0 SDK?

noahfalk replied:
Given that the OpenTelemetry crew might make breaking changes between pre-release builds it is probably safer to pin the exact version rather than allowing it to float via --prerelease. Once an RTM version is available I'd love to remove the '--version' from the docs entirely.

I did go ahead and add a warning that this is pre-release functionality, that breaking changes are possible, and that we only tested with this exact version.

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+
+    static void Main(string[] args)
+    {
+        using var meterProvider = Sdk.CreateMeterProviderBuilder()
```
>var [](http://example.com/codeflow?start=14&length=3)

would be better to use the actual type here instead of var.

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+        meterListener.Dispose();
+    }
+
+    static void OnMeasurementRecorded<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string,object>> tags, object? state)
```
>T [](http://example.com/codeflow?start=38&length=1)

should we make this <int>?

noahfalk replied:
I was doing all my examples using 3.1 SDK where adding the `?` for nullability would create compiler warnings and I was attempting to avoid additional explanatory text/code to work around it. In many areas I know our new features are often coupled with users adopting new versions of the SDK/runtime but given our broad back-compat on DiagnosticSource I am guessing that the majority of users for this API are not yet using C# nullability analysis.

If you think it is likely to cause confusion we could still add some text about it though. Let me know your thoughts.

noahfalk replied:
I realized the `object? state` was already causing a nullability warning a missed somehow so I removed it for now to be consistent on not provided nullability markup.

tarekgh replied:
Oh, my comment was not specific to nullability. I see my comment was malformed. sorry about that. My comment is to use `int` instead of `T` on this callback method? just to educate users using specific type would have a better perf. But I am ok too if you think using `T` is simpler for this example.

noahfalk replied:
Ah no worries Tarek. I'll stick with the T for now because I think it will let users extend the scenario more easily. I haven't measured, but as best I understand this isn't a case where the generics are creating perf overhead.

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+
+    static void Main(string[] args)
+    {
+        MeterListener meterListener = new MeterListener();
```
>MeterListener meterListener = new MeterListener(); [](http://example.com/codeflow?start=8&length=50)

maybe can use 

```C#
using MeterListener meterListener = new MeterListener();
```

and remove `meterListener.Dispose();` at the end?

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+
+Lets break down what happens in the example above. 
+```C#
+using var meterListener = new MeterListener();
````
>using [](http://example.com/codeflow?start=0&length=5)

interesting, you already using it here :-)

noahfalk replied:
Yeah I was switching back and forth between the two approaches trying to get a better flow for the explanatory text. I wound up with inconsistent text : ) It should be consistent now as `using MeterListener meterListener = ...`

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+```C#
+meterListener.SetMeasurementEventCallback<int>(OnMeasurementRecorded);
+...
+static void OnMeasurementRecorded<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string,object>> tags, object? state)
````
>T [](http://example.com/codeflow?start=34&length=1)

If you decide to change this to `int` in the sample, you'll need to change it here too.

noahfalk replied:
I wanted to leave it as T because so that it easier for users to extend the sample with multiple calls to SetMeasurementEventCallback() on different data types. I expect only needing a single data type will be uncommon.

tarekgh replied:
This is reply to my other comment :-) don't you think it would be better for the perf to use non generic type? I am ok to leave it as `T` but add a small comment the user can provide non generic type too which would be perform better?

noahfalk replied:
> don't you think it would be better for the perf to use non generic type?

In this case, no I don't expect there is a perf difference but I'll admit I didn't do benchmarking to confirm that I didn't overlook something. I'm not planning to go benchmark it now, but if someone else did that and it turns out I had it wrong I'm happy to change it.

tarekgh commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+will also trigger InstrumentPublished to be created.
+
+```C#
+meterListener.Dispose();
````
you have used `using ` when explaining the code here. we need to ensure we are consistent if we are using `using` or we manually calling `Dispose`

noahfalk replied:
The code and explanation should now be consistent on the 'using' pattern.

tarekgh commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,452 @@
+- **Storage of aggregated data** - Most useful metrics need data to be aggregated from many measurements. One option is the caller provides individual measurements
+  at arbitrary times and the collection tool will manage the aggregation. Alternately the caller can manage the aggregate measurements and provide them on-demand in a callback.
+
+Types of instruments currently available:
```
would it be good to give some examples when each instrument can be used. For example, Counter can be used for counting number of requests. ObservableGauge can be used to report room temperature,...etc.

noahfalk replied:
There is a section lower in the text that is 'Best practices selecting an instrument type' which does have this info. Do you think I should restructure to bring the examples closer, repeat the examples, or something else?

noahfalk replied:
@tarekgh - making sure you see the question here for you : )

tarekgh replied:
I suggest bringing the section `Best practices selecting an instrument type` closer to this section. As the article reader, will give immediate idea why we have such instruments and when to use each one. I am ok if you feel otherwise too. this is just a suggestion :-)

reyang replied:
FYI - we have some guidance regarding which instrument to be used in the OpenTelemetry spec repo https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/supplementary-guidelines.md#instrument-selection.

tarekgh commented:
@noahfalk this looks great to me.

reyang commented on docs/core/diagnostics/index.md:
```diff
@@ -21,7 +21,9 @@ This article helps you find the various tools you need.
 
-[EventCounters](event-counters.md) allows you to write metrics to identify and monitor performance issues. Metrics incur lower performance overhead compared to tracing, making it more suitable for an always-on performance monitoring. The .NET runtime and libraries publish several [well-known EventCounters](available-counters.md) that you can monitor as well.
+[Metrics](metrics.md) are numerical measurements recorded by apps and libraries over time, often to monitor application health and generate alerts when potential
+problems are detected. Metrics incur lower performance overhead compared to tracing, making them more suitable for an always-on monitoring.
```
Minor: What does "tracing" mean here? I wonder if folks would mix "tracing" with "logs with `dotnet-trace`"?

noahfalk replied:
yeah the terminology is ambiguous because people use 'tracing' both as shorthand for 'distributed tracing' and
to mean high verbosity logging. I rephrased the message to avoid refering to logging/tracing at all. I think the most
important information is that metrics are cheap and intended for always-on use. Their cost relative to other types of
telemetry feels more nuanced and probably wasn't necessary for the overview text. Let me know if this sounds better.

(There is probably still ambiguity in other parts of the overview doc, like the Logging and Tracing section immediately above, but scoping my changes to just the metrics portion of the text for now)

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+title: Collect metrics - .NET
+description: Tutorial to collect metrics in .NET applications
+ms.topic: tutorial
+ms.date: 10/27you/2021
```
```suggestion
ms.date: 10/27/2021
```

noahfalk replied:
that is interesting, not sure how that got there : )

reyang commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+
+```dotnetcli
+> dotnet tool update -g dotnet-counters
+You can invoke the tool using the following command: dotnet-counters
````
I guess this line is intended to be outside a code block, currently it seems to be inside a code block by accident?

noahfalk replied:
That text "You can invoke the tool using the following command: dotnet-counters" is what is printed on the command-line output when you run the command so I think it is formatted correctly. Let me know if I misunderstand.

![image](https://user-images.githubusercontent.com/6243776/140670262-32793268-0c71-4738-b2ce-e3f51186f729.png)

reyang replied:
Please resolve this comment (I don't seem to have permission to do that).

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+This example shows one of the integrations available now for OpenTelemetry metrics using the very popular OSS 
+[Prometheus](https://prometheus.io/) and [Grafana](https://grafana.com/) projects. The metrics data will flow like this:
+
+1. The .NET metric APIs collect measurements from our sample application
```
```suggestion
1. The .NET metric APIs collect measurements from our sample application.
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+[Prometheus](https://prometheus.io/) and [Grafana](https://grafana.com/) projects. The metrics data will flow like this:
+
+1. The .NET metric APIs collect measurements from our sample application
+2. The OpenTelemetry library running inside the same process aggregates these measurements
```
```suggestion
2. The OpenTelemetry library running inside the same process aggregates these measurements.
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+2. The OpenTelemetry library running inside the same process aggregates these measurements
+3. The Prometheus exporter library makes the aggregated data available via an HTTP
+metrics endpoint. 'Exporter' is what OpenTelemetry calls the libraries that transmit
+telemetry to vendor-specific services.
```
"backends" might be a better term https://github.com/open-telemetry/opentelemetry-specification/blob/16da729707e02e83f3a1b25b4faa3094c3f9e29b/specification/library-guidelines.md#sdk-implementation

```suggestion
telemetry to vendor-specific backends.
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+3. The Prometheus exporter library makes the aggregated data available via an HTTP
+metrics endpoint. 'Exporter' is what OpenTelemetry calls the libraries that transmit
+telemetry to vendor-specific services.
+4. A Prometheus service, potentially running on a different machine, polls the
```
Consider align with the Prometheus term and have a link:

https://prometheus.io/docs/introduction/first_steps/

"A Prometheus server, potentially running on a different machine, collects the metrics by scraping the metrics endpoint, reads the data ..."

noahfalk replied:
I put both the polling and scraping terminology now. Hopefully this is understandable for developers regardless whether they are experienced with Prometheus or not.

reyang replied:
Makes sense, thanks!

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+> The Promethus exporter library includes a reference to OpenTelemetry's shared library so this command implicitly adds both libraries
+> to the application.
+
+2. Modify the code of Program.cs so that it contains the extra code to configure OpenTelemetry at the beginning of Main():
```
```suggestion
2. Modify the code of `Program.cs` so that it contains the extra code to configure OpenTelemetry at the beginning of Main():
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+1. Follow the [Prometheus first steps](https://prometheus.io/docs/introduction/first_steps/) to set up your Prometheus server
+and confirm it is working.
+
+2. Modify the prometheus.yml configuration file so that Prometheus will scrape the metrics endpoint that our example app is
```
I noticed that we have both "example" and "sample" used, I suggest that we use "example" (sample/sampling/sampler have special meaning in the overall telemetry space).

reyang commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+2. Modify the prometheus.yml configuration file so that Prometheus will scrape the metrics endpoint that our example app is
+exposing. Add this text in the scrape_configs section:
+
+```
````
```suggestion
```yaml
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+      - targets: ['localhost:9184']
+```
+
+Reload the configuration or restart the prometheus server, then confirm that OpenTelemetryTest is in the UP
````
```suggestion
Reload the configuration or restart the Prometheus server, then confirm that OpenTelemetryTest is in the UP
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+[hats-sold-graph](media\prometheus-hat-sold-metric.png)
+
+If the Prometheus server hasn't been scraping (Prometheus' terminology for polling) the sample app for long you may need to wait a short
+while for data to accumulate. The default configuration has Prometheus scrape metric data every 15 seconds.
```
I suggest putting an explicit scraping period in the configuration file (e.g. 1 second) for better demo/learning experience.

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+1. Follow [the standard instructions](https://prometheus.io/docs/visualization/grafana/#creating-a-prometheus-graph) to install Grafana and
+connect it to a Prometheus data source.
+
+2. Create a Grafana dashboard by clicking the + icon on the left toolbar in the Grafana web portal, then select 'Dashboard'. In dashboard
```
```suggestion
2. Create a Grafana dashboard by clicking the "+" icon on the left toolbar in the Grafana web portal, then select 'Dashboard'. In dashboard
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+[hats-sold-dashboard-editor](media\grafana-hats-sold-dashboard-editor.png)
+
+3. Click Apply to save and view our simple new dashboard.
+[hats-sold-dashboard](media\grafana-hats-sold-dashboard)
```
```suggestion
[hats-sold-dashboard](media\grafana-hats-sold-dashboard.png)
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+
+3. Click Apply to save and view our simple new dashboard.
+[hats-sold-dashboard](media\grafana-hats-sold-dashboard)
+
```
nit: remove the extra white line.

reyang commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,404 @@
+being recorded by <xref:System.Diagnostics.Metrics.Meter?displayProperty=nameWithType>. For guidance creating custom
+logic compatible with the older EventCounters instrumentation see [here](event-counters.md).
+
+Modify the code of Program.cs to use the MeterListener like this:
```
```suggestion
Modify the code of `Program.cs` to use `MeterListener` like this:
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+...
+```
+
+Lets break down what happens in the example above. 
````
```suggestion
Let's break down what happens in the example above. 
```

reyang commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+    }
+};
+```
+Here we configured which instruments the listener will receive measurements from.
````
nit: add a white line above

reyang commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,404 @@
+    Console.WriteLine($"{instrument.Name} recorded measurement {measurement}");
+}
+```
+Next we configured the delegate that is invoked when measurements are received from an instrument by calling
````
nit: add an empty line above

reyang commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,452 @@
+
+### Types of Instruments
+
+Above the example created a <xref:System.Diagnostics.Metrics.Counter> instrument but there are several different types available. Instruments differ in two ways:
```
???

```suggestion
The above example created a <xref:System.Diagnostics.Metrics.Counter> instrument but there are several different types available. Instruments differ in two ways:
```

noahfalk replied:
Your suggestion did capture my intended meaning. Not sure if your "???" meant there was something else you found confusing as well?

reyang replied:
I found it hard to understand (perhaps due to my bad English). I interpreted it as "In the example we've only demonstrated the <xref:System.Diagnostics.Metrics.Counter> instrument, but there are more instrument types available."

noahfalk replied:
Your English is pretty sharp if you ask me ; ) Your phrasing was clearer so I adopted it.

CodeBlanch commented on docs/core/diagnostics/index.md:
```diff
@@ -21,7 +21,9 @@ This article helps you find the various tools you need.
 ## Metrics
 
-[EventCounters](event-counters.md) allows you to write metrics to identify and monitor performance issues. Metrics incur lower performance overhead compared to tracing, making it more suitable for an always-on performance monitoring. The .NET runtime and libraries publish several [well-known EventCounters](available-counters.md) that you can monitor as well.
+[Metrics](metrics.md) are numerical measurements recorded by apps and libraries over time, often to monitor application health and generate alerts when potential
```
Something about mixing "app" and "application" in the same sentence seems funny to me.

Here's my stab at wordsmithing it:

Metrics are numerical measurements taken while code is running and analyzed over time to monitor the health and/or performance of an application or system. Metrics are often used to generate alerts when potential problems are detected.

🤷

noahfalk replied:
I did some more tweaking and got here. How does this sound to you?

`Metrics are numerical measurements recorded over time to monitor application performance and health. Metrics are often used to generate alerts when potential problems are detected.`

reyang commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,452 @@
+  extremely hot code paths where performance is important and using <xref:System.Diagnostics.Metrics.Counter.Add%2A> would create >1 million calls/sec/thread, using
+  ObservableCounter may offer more opportunity for optimization.
+
+- For timing things Histogram is usually prefered. Often it is useful to understand the tail of these distributions (90th, 95th, 99th percentile) rather than averages or
```
```suggestion
- For timing things Histogram is usually preferred. Often it is useful to understand the tail of these distributions (90th, 95th, 99th percentile) rather than averages or
```

reyang commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,452 @@
+- Other common cases such as business metrics, physical sensors, cache hit rates, or sizes of caches, queues, files, etc are usually well suited for ObservableGauge.
+
+- Be aware that Histograms tend to store a lot more data in memory than other metric types however the exact memory usage is determined by the collection tool being used.
+  If you are defining a large number of Histogram metrics (>100) you may need to give users guidance not to enable them all at the same time or configure their tools to save
```
```suggestion
  If you are defining a large number (>100) of Histogram metrics you may need to give users guidance not to enable them all at the same time or configure their tools to save
```

reyang commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,452 @@
+### Descriptions and Units
+
+Instruments can specify optional descriptions and units. These values are opaque to all metric calculations but can be shown in collection tool UI
+to help engineers understand how to interpret the data. Stop the example process from above and replace the example code in Program.cs with:
```
```suggestion
to help engineers understand how to interpret the data. Stop the example process from above and replace the example code in `Program.cs` with:
```

reyang commented on docs/core/diagnostics/metrics.md:
```diff
@@ -0,0 +1,34 @@
+- [Collection tutorial](.\metrics-collection.md) - How to store and view metric data for your app
+- [Built-in metrics](.\available-counters.md) - Discover metrics that are ready for use in .NET runtime libraries
+
+
```
nit: remove these extra empty lines.

reyang reviewed (approved):
LGTM overall.

noahfalk committed:
PR feedback

noahfalk committed:
A few updates from re-reading past feedback/discussions with Juan

noahfalk committed:
fix lint/build issues

noahfalk committed:
Redo some changes that got accidentally overwritten

noahfalk committed:
Fix inline image links

noahfalk commented:
Thanks for all the review feedback. Most of the feedback I commented on now has changes present but I noticed that I messed up some of the changes and reverted or overwrote them accidentally. I am fixing it, but need to sleep for now. I'll ping again as soon I get it all fixed.

noahfalk committed:
More PR feedback changes

noahfalk commented:
OK, I think I've got all the changes in place for feedback so far and I removed all the WIP markers. If there is more feedback I'm glad to address it.

@dotnet/docs - could someone take a look at this? Thanks!

noahfalk commented:
@Anipik - sorry I should have added you earlier. You are welcome to look at anything but in particular I wanted to see if you had any concerns about the statements on the PerformanceCounter APIs in the API comparison.

cijothomas commented on docs/core/diagnostics/understanding-different-metric-apis.md:
```diff
@@ -0,0 +1,74 @@
+the metrics API directly into the base class libraries to eliminate any friction normally associated with .NET libraries adding 3rd party
+dependencies. It is available either by targetting .NET 6 or older .NET Core and .NET Framework apps can add a reference to the .NET
+[System.Diagnostics.DiagnosticsSource](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource) 6.0 NuGet package. In addition to
+aiming at broad compatibility this API adds support for many things that were lacking from EventCounters such as histograms
```
```suggestion
aiming at broad compatibility, this API adds support for many things that were lacking from EventCounters such as histograms
```

cijothomas commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,462 @@
+  frequently prefer using the same sequence of tag names for each call.
+
+- The .NET API is optimized to be allocation-free for <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A> calls
+  with three or fewer tags. In general the performance overhead of these calls increases as more tags are used.
```
mention `TagList` for alloc free above 3 as well.

tarekgh replied:
TagList can help in avoid the allocation if having up to 8 tags :-) will have small cpu overhead for copying the tags.

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+
+- Populating metrics in Grafana with OpenTelemetry and Prometheus
+- Viewing metrics in real-time with the dotnet-counters command-line tool
+- Creating a custom collection tool using the underlying .NET <xref:System.Diagnostics.Metrics.MeterListener> API
```
```suggestion
- Creating a custom collection tool using the underlying .NET <xref:System.Diagnostics.Metrics.MeterListener> API.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+- Creating a custom collection tool using the underlying .NET <xref:System.Diagnostics.Metrics.MeterListener> API
+
+For more information on custom metric instrumentation see [here](understanding-different-metric-apis.md) for an overview
+of instrumentation options.
```
```suggestion
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+[dotnet-counters](dotnet-counters.md) is a simple command-line tool that can view live metrics for any .NET Core application on
+demand. It doesn't require any advance setup which can make it very useful for ad-hoc investigations or to verify that metric
+instrumentation is working correctly. It works with both <xref:System.Diagnostics.Metrics?displayProperty=nameWithType> based
+APIs as well as [EventCounters](event-counters.md).
```
```suggestion
APIs and [EventCounters](event-counters.md).
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+instrumentation is working correctly. It works with both <xref:System.Diagnostics.Metrics?displayProperty=nameWithType> based
+APIs as well as [EventCounters](event-counters.md).
+
+If the [dotnet-counters](dotnet-counters.md) tool is not already installed use the SDK to install it:
```
```suggestion
If the [dotnet-counters](dotnet-counters.md) tool is not already installed, use the SDK to install it:
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+Tool 'dotnet-counters' (version '5.0.251802') was successfully installed.
+```
+
+While the example app above is still running, list the running processes in a 2nd shell to determine the process ID:
````
```suggestion
While the example app is still running, list the running processes in a second shell to determine the process ID:
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+```
+
+Find the ID for the process name that matches the example app and have dotnet-counters monitor all metrics from the
+"HatCo.HatStore" Meter. Be careful, the meter name is case-sensitive.
````
```suggestion
"HatCo.HatStore" meter. The meter name is case-sensitive.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+    Working Set (MB)                                              30
+```
+
+See the [dotnet-counters](dotnet-counters.md) for more information about the tool or [builtin metrics](available-counters.md) to learn
````
```suggestion
For more information about the tool, see the [dotnet-counters](dotnet-counters.md).
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+```
+
+See the [dotnet-counters](dotnet-counters.md) for more information about the tool or [builtin metrics](available-counters.md) to learn
+more about metrics that are available out of the box in .NET.
````
```suggestion
To learn more about metrics that are available out of the box in .NET, see [built-in metrics](available-counters.md).
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+
+[OpenTelemetry](https://opentelemetry.io/) is a vendor-neutral open-source project supported by the
+[Cloud Native Computing Foundation](https://www.cncf.io/) that aims to standardize generating and collecting telemetry for
+cloud-native software. The .NET team has designed our built-in platform metric APIs to be compatible with this
```
```suggestion
cloud-native software. The built-in platform metric APIs are designed to be compatible with this
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+[OpenTelemetry](https://opentelemetry.io/) is a vendor-neutral open-source project supported by the
+[Cloud Native Computing Foundation](https://www.cncf.io/) that aims to standardize generating and collecting telemetry for
+cloud-native software. The .NET team has designed our built-in platform metric APIs to be compatible with this
+standard to make integration straightforward for any .NET developers that wish to use it. At the time of writing support for
```
```suggestion
standard to make integration straightforward for any .NET developers that wish to use it. At the time of writing, support for
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+[Cloud Native Computing Foundation](https://www.cncf.io/) that aims to standardize generating and collecting telemetry for
+cloud-native software. The .NET team has designed our built-in platform metric APIs to be compatible with this
+standard to make integration straightforward for any .NET developers that wish to use it. At the time of writing support for
+OpenTelemetry metrics is relatively new but [Azure Monitor](https://docs.microsoft.com/azure/azure-monitor/app/opentelemetry-overview)
```
```suggestion
OpenTelemetry metrics is relatively new, but [Azure Monitor](/azure/azure-monitor/app/opentelemetry-overview)
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+metrics endpoint. 'Exporter' is what OpenTelemetry calls the libraries that transmit
+telemetry to vendor-specific backends.
+4. A Prometheus server, potentially running on a different machine, polls the
+metrics endpoint, reads the data and stores it in a database for long-term persistence.
```
```suggestion
metrics endpoint, reads the data, and stores it in a database for long-term persistence.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+`AddMeter("HatCo.HatStore")` configures OpenTelemetry to transmit all the metrics collected by the Meter our app defined.
+`AddPrometheusExporter(...)` configures OpenTelemetry to expose Prometheus' metrics endpoint on port 9184 and to use
+the HttpListener. See the [OpenTelemetry documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.Prometheus)
+for more information about OpenTelemetry configuration options, in particular alternate hosting options that are useful for ASP.NET applications.
```
```suggestion
for more information about OpenTelemetry configuration options, in particular, alternative hosting options that are useful for ASP.NET applications.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+
+> [!NOTE]
+> At the time of writing OpenTelemetry only supports metrics emitted using the <xref:System.Diagnostics.Metrics?displayProperty=nameWithType>
+> APIs however support for [EventCounters](event-counters.md) is planned.
```
```suggestion
> APIs; however, support for [EventCounters](event-counters.md) is planned.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+Press any key to exit
+```
+
+#### Setup and Configure Prometheus
````
```suggestion
#### Set up and configure Prometheus
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+```
+
+Reload the configuration or restart the Prometheus server, then confirm that OpenTelemetryTest is in the UP
+state in the Status->Targets page of the Prometheus web portal.
````
```suggestion
state in the **Status** > **Targets** page of the Prometheus web portal.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+Reload the configuration or restart the Prometheus server, then confirm that OpenTelemetryTest is in the UP
+state in the Status->Targets page of the Prometheus web portal.
+
+On the Graph page of the Prometheus web portal enter `hats_sold` in the expression text box. In the graph tab Prometheus should
```
```suggestion
On the Graph page of the Prometheus web portal, enter `hats_sold` in the expression text box. In the graph tab, Prometheus should
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+show the steadily increasing value of the "hats-sold" Counter that is being emitted by our example application.
+[![Prometheus hats sold graph](media/prometheus-hat-sold-metric.png)](media/prometheus-hat-sold-metric.png)
+
+If the Prometheus server hasn't been scraping the example app for long you may need to wait a short while for data to accumulate.
```
```suggestion
If the Prometheus server hasn't been scraping the example app for long, you may need to wait a short while for data to accumulate.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+1. Follow [the standard instructions](https://prometheus.io/docs/visualization/grafana/#creating-a-prometheus-graph) to install Grafana and
+connect it to a Prometheus data source.
+
+2. Create a Grafana dashboard by clicking the "+" icon on the left toolbar in the Grafana web portal, then select 'Dashboard'. In the dashboard
```
```suggestion
2. Create a Grafana dashboard by clicking the **+** icon on the left toolbar in the Grafana web portal, then select **Dashboard**. In the dashboard
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+editor that appears enter 'Hats Sold/Sec' as the Title and 'rate(hats_sold[5m])' in the PromQL expression field. It should look like this:
+[![Hats sold Grafana dashboard editor](media/grafana-hats-sold-dashboard-editor.png)](media/grafana-hats-sold-dashboard-editor.png)
+
+3. Click Apply to save and view our simple new dashboard.
```
```suggestion
3. Click **Apply** to save and view the simple new dashboard.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+
+The .NET <xref:System.Diagnostics.Metrics.MeterListener> API allows creating custom in-process logic to observe the measurements
+being recorded by <xref:System.Diagnostics.Metrics.Meter?displayProperty=nameWithType>. For guidance creating custom
+logic compatible with the older EventCounters instrumentation see [here](event-counters.md).
```
```suggestion
logic compatible with the older EventCounters instrumentation, see [EventCounters](event-counters.md).
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+}
+```
+
+When run the application now runs our custom callback on each measurement:
````
```suggestion
When run, the application now runs our custom callback on each measurement:
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+using MeterListener meterListener = new MeterListener();
+```
+
+First we created an instance of the <xref:System.Diagnostics.Metrics.MeterListener> which we will use to receive measurements.
````
```suggestion
First we created an instance of the <xref:System.Diagnostics.Metrics.MeterListener>, which we will use to receive measurements.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+Here we configured which instruments the listener will receive measurements from.
+<xref:System.Diagnostics.Metrics.MeterListener.InstrumentPublished> is a delegate that will be invoked any time a new
+instrument is created within the app. Our delegate can examine the instrument, such as checking the name, the Meter, or any other
+public property to decide whether to subscribe. If we do want to receive measurements from this instrument then we invoke
```
```suggestion
public property to decide whether to subscribe. If we do want to receive measurements from this instrument, then we invoke
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+instrument is created within the app. Our delegate can examine the instrument, such as checking the name, the Meter, or any other
+public property to decide whether to subscribe. If we do want to receive measurements from this instrument then we invoke
+<xref:System.Diagnostics.Metrics.MeterListener.EnableMeasurementEvents%2A> to indicate that. If your code has another way
+to obtain a reference to an instrument it is legal to invoke EnableMeasurementEvents() at any time with that reference, but this is
```
```suggestion
to obtain a reference to an instrument, it's legal to invoke `EnableMeasurementEvents()` at any time with that reference, but this is
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+Next we configured the delegate that is invoked when measurements are received from an instrument by calling
+<xref:System.Diagnostics.Metrics.MeterListener.SetMeasurementEventCallback%2A>. The generic parameter controls which data type
+of measurement will be received by the callback. For example a `Counter<int>` generates `int` measurements whereas a
+`Counter<double>` will generate `double` measurements. Instruments are allowed to be created with `byte`, `short`, `int`, `long`,
```
```suggestion
`Counter<double>` generates `double` measurements. Instruments are allowed to be created with `byte`, `short`, `int`, `long`,
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+`float`, `double`, and `decimal` types. We recommend registering a callback for every data type unless you have scenario
+specific knowledge that not all data types will be needed, such as in this example. Making repeated calls to
+SetMeasurementEventCallback() with different generic arguments may appear a little unusual. The API is designed this way
+to allow MeterListeners to receive measurements with extremely low performance overhead, typically just a few ns on the
```
```suggestion
to allow MeterListeners to receive measurements with extremely low performance overhead, typically just a few nanoseconds.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+specific knowledge that not all data types will be needed, such as in this example. Making repeated calls to
+SetMeasurementEventCallback() with different generic arguments may appear a little unusual. The API is designed this way
+to allow MeterListeners to receive measurements with extremely low performance overhead, typically just a few ns on the
+recent .NET versions.
```
```suggestion
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+to allow MeterListeners to receive measurements with extremely low performance overhead, typically just a few ns on the
+recent .NET versions.
+
+When MeterListener.EnableMeasurementEvents() was called initially there was an oportunity to provide a `state` object as
```
```suggestion
When MeterListener.EnableMeasurementEvents() was called initially, there was an opportunity to provide a `state` object as
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+recent .NET versions.
+
+When MeterListener.EnableMeasurementEvents() was called initially there was an oportunity to provide a `state` object as
+one of the parameters and that object can be anything you want. If you provide a state object in that call then it will be
```
```suggestion
one of the parameters. That object can be anything you want. If you provide a state object in that call, then it will be
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+stored with that instrument and returned to you as the `state` parameter in the callback. This is intended both as a
+convenience and as a performance optimization. Often listeners need to create an object for each instrument that will
+store measurements in memory and have code to do calculations on those measurements. Although you could create a Dictionary
+that maps from the instrument to the storage object and look it up on every measurement that would be much slower than
```
```suggestion
that maps from the instrument to the storage object and look it up on every measurement, that would be much slower than
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+meterListener.Start();
+```
+
+Once the MeterListener is configured we need to start it which triggers callbacks to begin. The InstrumentPublished
````
```suggestion
Once the `MeterListener` is configured, we need to start it to trigger callbacks to begin. The `InstrumentPublished`
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+
+Once the MeterListener is configured we need to start it which triggers callbacks to begin. The InstrumentPublished
+delegate will be invoked for every pre-existing Instrument in the process. In the future any newly created Instrument
+will also trigger InstrumentPublished to be created.
```
```suggestion
will also trigger `InstrumentPublished` to be invoked.
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
````diff
@@ -0,0 +1,419 @@
+```
+
+Once the MeterListener is configured we need to start it which triggers callbacks to begin. The InstrumentPublished
+delegate will be invoked for every pre-existing Instrument in the process. In the future any newly created Instrument
````
```suggestion
delegate will be invoked for every pre-existing Instrument in the process. In the future, any newly created Instrument
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+
+Once we are done listening, disposing the listener stops the flow of callbacks and releases any internal references
+to the listener object. The `using` keyword we used when declaring `meterListener` causes Dispose() to be called automatically
+when the variable goes out of scope. Be wary that Dispose() is only promising that it won't initiate new callbacks. Because callbacks
```
```suggestion
when the variable goes out of scope. Be aware that `Dispose()` is only promising that it won't initiate new callbacks. Because callbacks
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+Once we are done listening, disposing the listener stops the flow of callbacks and releases any internal references
+to the listener object. The `using` keyword we used when declaring `meterListener` causes Dispose() to be called automatically
+when the variable goes out of scope. Be wary that Dispose() is only promising that it won't initiate new callbacks. Because callbacks
+occur on different threads there may still be callbacks in progress after the call to Dispose() returns. If you need a
```
```suggestion
occur on different threads, there may still be callbacks in progress after the call to `Dispose()` returns. If you need a
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+to the listener object. The `using` keyword we used when declaring `meterListener` causes Dispose() to be called automatically
+when the variable goes out of scope. Be wary that Dispose() is only promising that it won't initiate new callbacks. Because callbacks
+occur on different threads there may still be callbacks in progress after the call to Dispose() returns. If you need a
+guarantee that a certain region of code in your callback isn't currently executing and will never execute again in the future
```
```suggestion
guarantee that a certain region of code in your callback isn't currently executing and will never execute again in the future,
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+when the variable goes out of scope. Be wary that Dispose() is only promising that it won't initiate new callbacks. Because callbacks
+occur on different threads there may still be callbacks in progress after the call to Dispose() returns. If you need a
+guarantee that a certain region of code in your callback isn't currently executing and will never execute again in the future
+then you will need some additional thread synchronization to enforce that. Dispose() doesn't include the synchronization
```
```suggestion
then you will need some additional thread synchronization to enforce that. `Dispose()` doesn't include the synchronization
```

gewarren commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+occur on different threads there may still be callbacks in progress after the call to Dispose() returns. If you need a
+guarantee that a certain region of code in your callback isn't currently executing and will never execute again in the future
+then you will need some additional thread synchronization to enforce that. Dispose() doesn't include the synchronization
+by default because it adds performance overhead in every measurement callback and MeterListener is designed as a very
```
```suggestion
by default because it adds performance overhead in every measurement callback&mdash;and `MeterListener` is designed as a highly
```

gewarren reviewed (commented):
I didn't review metrics.md, understanding-different-metric-apis.md, or metrics-instrumentation.md yet, but wanted to pass along these suggestions meanwhile.

noahfalk committed:
PR feedback

noahfalk committed:
Update toc

noahfalk commented on docs/core/diagnostics/metrics-collection.md:
```diff
@@ -0,0 +1,419 @@
+---
```
Thanks! I updated the TOC so hopefully everything is covered now.

noahfalk committed:
Move instrument selection best practices and add TagList

noahfalk commented:
Thanks for the great feedback everyone! I've attempted to incorporate all of it in the most recent changes.

noahfalk commented:
Does anyone need more time to review further or should I go ahead and merge? Thanks all!

reyang commented:
> Does anyone need more time to review further or should I go ahead and merge? Thanks all!

:shipit:

gewarren commented:
> Does anyone need more time to review further or should I go ahead and merge? Thanks all!

I'll take another look in the next couple hours - thanks!

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+### Multi-dimensional metrics
+
+Measurements can also be associated with key-value pairs called tags that allow data to be categorized for analysis. For example HatCo might want to record not
+only the number of hats that were sold, but also which size and color they were. Later when analyzing the data HatCo engineers can break out the totals by
```
```suggestion
only the number of hats that were sold, but also which size and color they were. When analyzing the data later, HatCo engineers can break out the totals by
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+only the number of hats that were sold, but also which size and color they were. Later when analyzing the data HatCo engineers can break out the totals by
+size, color, or any combination of both.
+
+Counter and Histogram tags can be specified in overloads of the <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and
```
```suggestion
Counter and Histogram tags can be specified in overloads of the <xref:System.Diagnostics.Metrics.Counter%601.Add%2A> and
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+size, color, or any combination of both.
+
+Counter and Histogram tags can be specified in overloads of the <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and
+<xref:System.Diagnostics.Metrics.Histogram`1.Record%2A> respectively that take one or more KeyValuePairs. For example:
```
```suggestion
<xref:System.Diagnostics.Metrics.Histogram%601.Record%2A> that take one or more `KeyValuePair` arguments. For example:
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
````diff
@@ -0,0 +1,463 @@
+
+```
+
+Replace the code of `Program.cs` and re-run the app and dotnet-counters as before:
````
```suggestion
Replace the code of `Program.cs` and rerun the app and dotnet-counters as before:
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+- Beware of having very large or unbounded combinations of tag values being recorded in practice. Although the .NET API implementation can handle it, collection tools will
+  likely allocate storage for metric data associated with each tag combination and this could become very large. For example it is fine if HatCo has 10 different
+  hat colors and 25 hat sizes for up to 10*25=250 sales totals to track. However if HatCo added a 3rd tag that is a CustomerID for the sale and they sell to 100
+  million customers world-wide now there are now likely to be billions of different tag combinations being recorded. Most metric collection tools will either drop data
```
```suggestion
  million customers worldwide, now there are now likely to be billions of different tag combinations being recorded. Most metric collection tools will either drop data
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  hat colors and 25 hat sizes for up to 10*25=250 sales totals to track. However if HatCo added a 3rd tag that is a CustomerID for the sale and they sell to 100
+  million customers world-wide now there are now likely to be billions of different tag combinations being recorded. Most metric collection tools will either drop data
+  to stay within technical limits or there can be large monetary costs to cover the data storage and processing. The implementation of each collection tool will determine
+  its limits but likely < 1K combinations for one instrument is safe, above 1K will need the collection tool to apply filtering or be engineered to operate at high scale.
```
```suggestion
  its limits, but likely less than 1000 combinations for one instrument is safe. Anything above 1000 combinations will require the collection tool to apply filtering or be engineered to operate at high scale.
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  million customers world-wide now there are now likely to be billions of different tag combinations being recorded. Most metric collection tools will either drop data
+  to stay within technical limits or there can be large monetary costs to cover the data storage and processing. The implementation of each collection tool will determine
+  its limits but likely < 1K combinations for one instrument is safe, above 1K will need the collection tool to apply filtering or be engineered to operate at high scale.
+  Histogram implementations tend to use far more memory than other metrics so safe limits could be 10-100x lower. If you anticipate large number of unique tag combinations
```
```suggestion
  Histogram implementations tend to use far more memory than other metrics, so safe limits could be 10-100 times lower. If you anticipate large number of unique tag combinations,
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  to stay within technical limits or there can be large monetary costs to cover the data storage and processing. The implementation of each collection tool will determine
+  its limits but likely < 1K combinations for one instrument is safe, above 1K will need the collection tool to apply filtering or be engineered to operate at high scale.
+  Histogram implementations tend to use far more memory than other metrics so safe limits could be 10-100x lower. If you anticipate large number of unique tag combinations
+  then logs, transactional databases or big data processing systems may be more appropriate solutions to operate at the needed scale.
```
```suggestion
  then logs, transactional databases, or big data processing systems may be more appropriate solutions to operate at the needed scale.
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  Histogram implementations tend to use far more memory than other metrics so safe limits could be 10-100x lower. If you anticipate large number of unique tag combinations
+  then logs, transactional databases or big data processing systems may be more appropriate solutions to operate at the needed scale.
+
+- For instruments that will have very large numbers of tag combinations, prefer using a smaller storage type to help reduce memory overhead. For example storing the `short` for
```
```suggestion
- For instruments that will have very large numbers of tag combinations, prefer using a smaller storage type to help reduce memory overhead. For example, storing the `short` for
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  then logs, transactional databases or big data processing systems may be more appropriate solutions to operate at the needed scale.
+
+- For instruments that will have very large numbers of tag combinations, prefer using a smaller storage type to help reduce memory overhead. For example storing the `short` for
+  a `Counter<short>` only occupies 2 bytes per tag combination whereas a `double` for `Counter<double>` occupies 8 bytes per tag combination.
```
```suggestion
  a `Counter<short>` only occupies 2 bytes per tag combination, whereas a `double` for `Counter<double>` occupies 8 bytes per tag combination.
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  a `Counter<short>` only occupies 2 bytes per tag combination whereas a `double` for `Counter<double>` occupies 8 bytes per tag combination.
+
+- Collection tools are encouraged to optimize for code that specifies the same set of tag names in the same order for each call to record measurements on the
+  same instrument. For high performance code that needs to call <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A>
```
```suggestion
  same instrument. For high performance code that needs to call <xref:System.Diagnostics.Metrics.Counter%601.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram%601.Record%2A>
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+
+- Collection tools are encouraged to optimize for code that specifies the same set of tag names in the same order for each call to record measurements on the
+  same instrument. For high performance code that needs to call <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A>
+  frequently prefer using the same sequence of tag names for each call.
```
```suggestion
  frequently, prefer using the same sequence of tag names for each call.
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  same instrument. For high performance code that needs to call <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A>
+  frequently prefer using the same sequence of tag names for each call.
+
+- The .NET API is optimized to be allocation-free for <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A> calls
```
```suggestion
- The .NET API is optimized to be allocation-free for <xref:System.Diagnostics.Metrics.Counter%601.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram%601.Record%2A> calls
```

gewarren commented on docs/core/diagnostics/metrics-instrumentation.md:
```diff
@@ -0,0 +1,463 @@
+  frequently prefer using the same sequence of tag names for each call.
+
+- The .NET API is optimized to be allocation-free for <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A> calls
+  with three or fewer tags specified individually. To avoid allocations with larger numbers of tags, use the <xref:System.Diagnostics.TagList>. In general,
```
```suggestion
  with three or fewer tags specified individually. To avoid allocations with larger numbers of tags, use <xref:System.Diagnostics.TagList>. In general,
```

gewarren reviewed (commented):
Left a bunch more comments. If you can batch these into a single commit (you can do that on the Files changed tab in GitHub), that would be great. Otherwise, an email is sent and a build kicked off for each commit.

I want to check the preview again after these changes are made to see how the headings look.

noahfalk committed:
Apply suggestions from code review

All the batchable changes here, a little more still to come that I need to handle manually

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

noahfalk committed:
Add links to other metrics pages

noahfalk committed:
Fix lint error

noahfalk commented:
Thanks for the thorough review @gewarren : ) All the changes you suggested should be reflected now.

gewarren committed:
reorder toc; fix some headings

gewarren committed:
acrolinx

gewarren commented:
@noahfalk I hope you don't mind that I made some direct commits. It was much easier to work with the articles holistically than line-by-line in the GitHub editor. I reordered the TOC a bit (nested the reference article under EventCounters) and renamed "Understanding metric APIs" to "Compare metric APIs". I also ran Acrolinx (our grammar checker).

gewarren commented on docs/core/diagnostics/available-counters.md:
```diff
@@ -7,7 +7,7 @@ ms.date: 12/17/2020
 # Well-known EventCounters in .NET
 
-The .NET runtime and libraries implement and publish several [`EventCounter`](./event-counters.md) that can be used to identify and diagnose various performance issues. This document is a reference on the providers that can be used to monitor these `EventCounters` and their descriptions.
+The .NET runtime and libraries implement and publish several [`EventCounter`](./event-counters.md) counters that can be used to identify and diagnose various performance issues. This document is a reference on the providers that can be used to monitor these `EventCounter` counters and their descriptions.
```
Is EventCounter actually a type? Or should we just refer to them as event counters and remove all instances of "EventCounter"?

gewarren committed:
change EventCounters to event counters

gewarren committed:
another eventcounters change

noahfalk commented:
@gewarren - The first two commits look great but I am a little concerned about renaming EventCounters. We have used "EventCounter" as both the name of a specific type (System.Diagnostics.Tracing.EventCounter) and as the name of the overall feature, a group of related APIs (EventCounter, PollingCounter, IncrementingEventCounter, IncrementingPollingCounter). It feels odd to retroactively change the naming from EventCounter -> "event counter".

gewarren committed:
move back to EventCounters

gewarren commented:
> @gewarren - The first two commits look great but I am a little concerned about renaming EventCounters. We have used "EventCounter" as both the name of a specific type (System.Diagnostics.Tracing.EventCounter) and as the name of the overall feature, a group of related APIs (EventCounter, PollingCounter, IncrementingEventCounter, IncrementingPollingCounter). It feels odd to retroactively change the naming from EventCounter -> "event counter".

Yes, after reading the "compare metrics APIs" article again, I realized I should change it back too. It was just odd to me to use an upper case name that looks like a type to refer to something bigger than a specific type.

I think the only remaining thing for me is that there is a section ".NET diagnostics tutorials" that has, for example, a tutorial on EventCounters. Should we move that tutorial under the Metrics parent instead? Or should we move the new collection and instrumentation tutorials under that Tutorials node?

![image](https://user-images.githubusercontent.com/24882762/141585441-8743fe9c-c536-4152-9869-8ebd3809a984.png)

noahfalk commented:
> I think the only remaining thing for me is that there is a section ".NET diagnostics tutorials" that has, for example, a tutorial on EventCounters. Should we move that tutorial under the Metrics parent instead? 

That makes sense to me. I think the original intent of that ".NET diagnostics tutorials" was to offer troubleshooting guides for different types of issues, but over time a few other tutorials have slipped in that aren't troubleshooting in nature.

noahfalk commented:
> It was just odd to me to use an upper case name that looks like a type to refer to something bigger than a specific type

Yeah the naming evolved organically rather than being explicitly decided. In .NET 4.7.1 we created EventCounter the type, so EventCounters at that time referred to having multiple instances of the type. Later we created some closely related types (PollingCounter, IncrementingEventCounter, etc) but continued to refer to the group of them as 'EventCounters'. At that point 'EventCounters' implicitly became a proper noun referring to a feature as a whole. For example people would say "The EventCounters feature helps developers monitor their applications", or "I suggest you use the EventCounters API for that job." 

I'm entirely trying to find a definition that matches pre-existing usage, but hopefully it seems workable? Thanks!

gewarren committed:
move eventcounters tutorial in toc

gewarren committed:
revert event counters change

gewarren committed:
readd changes that were reverted not related to naming

gewarren commented:
Merged! 🚀
